drivers/wdm/audio/legacy/wdmaud/mmixer.c

```diff
@@ -769,6 +769,27 @@ WdmAudMidiCapabilities(
         return STATUS_UNSUCCESSFUL;
 }
 
+NTSTATUS
+NTAPI
+WdmAudGetPosition(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PIRP Irp,
+    _In_ PWDMAUD_DEVICE_INFO DeviceInfo)
+{
+    MIXER_STATUS Status;
+    DWORD Position;
+
+    /* get position */
+    Status = MMixerGetWavePosition(&MixerContext, DeviceInfo->hDevice, &Position);
+
+    DeviceInfo->u.Position = (ULONGLONG)Position;
```

Comment:
Shouldn't the position change in DeviceInfo be done only if the GetWavePosition returned success?

Reply:
Yep, it should. But it's already done in `MMixerGetWavePosition` function. So need I do it here, if on failure, that function already returns zero position? :thinking:

Reply:
One would have to look at the code of this function to see that it returns a "valid" (zero) position even on failure. I argue this is not evident and usually expected (see all the e.g. CreateFile and other functions that return a valid value by pointer only in success, and otherwise the returned value is _**undefined**_ ).

Reply:
OK @HBelusca. Just checked: `MixerContext->Control` is a wrapper around `KsSynchronousIoControlDevice` in case of Legacy mode, which uses wdmaud.sys kernel-mode routines. This function, in its turn, calls `IoBuildDeviceIoControlRequest` and `IoCallDriver`, which calls a major function with an appropriate index. The returned value in failure case seems indeed undefined, so I changed this code as you proposed: store received position only in success case. It looks much more logically correct: by default, `DeviceInfo`'s position member has zero value, and in failure case it should remain such, without changing it to undefined value. :slightly_smiling_face: